File: RecompSwift/RefactorKit/Sources/Utilities/RicoApplyResult.swift
import Foundation
import os.log

public struct RicoSkippedAction: Sendable {
    public let type: String
    public let reason: String

    public init(type: String, reason: String) {
        self.type = type
        self.reason = reason
    }
}

public struct RicoApplyResult: Sendable {
    public var applied: [String] = []
    public var skipped: [RicoSkippedAction] = []
    public var touchedMeals = false
    public var touchedPlan = false
    public var shouldRegeneratePlan = false
    public var pendingRegenerateOptions = RegeneratePlanOptions.default

    public var hasLocalChanges: Bool { touchedMeals || touchedPlan }

    public var statusSuffix: String? {
        var parts: [String] = []
        if !applied.isEmpty {
            parts.append("Applied \(applied.count) change(s).")
        }
        if !skipped.isEmpty {
            let detail = skipped.map { "\($0.type) (\($0.reason))" }.joined(separator: "; ")
            parts.append("Ref couldn't apply: \(detail)")
        }
        guard !parts.isEmpty else { return nil }
        return "\n\n" + parts.joined(separator: " ")
    }

    public mutating func recordApplied(_ type: String) {
        applied.append(type)
    }

    public mutating func recordSkipped(type: String, reason: String) {
        skipped.append(RicoSkippedAction(type: type, reason: reason))
    }
}

File: recomp-mobile/app/src/main/java/com/refactor/app/api/RicoApplyResult.kt
package com.refactor.app.api

data class RicoSkippedAction(
    val type: String,
    val reason: String,
)

data class RicoApplyResult(
    val applied: List<String> = emptyList(),
    val skipped: List<RicoSkippedAction> = emptyList(),
    val touchedMeals: Boolean = false,
    val touchedPlan: Boolean = false,
) {
    fun recordApplied(type: String): RicoApplyResult =
        copy(applied = applied + type)

    fun recordSkipped(type: String, reason: String): RicoApplyResult =
        copy(skipped = skipped + RicoSkippedAction(type, reason))

    fun statusSuffix(): String {
        val parts = mutableListOf<String>()
        if (applied.isNotEmpty()) {
            parts += "Applied ${applied.size} change(s)."
        }
        if (skipped.isNotEmpty()) {
            val detail = skipped.joinToString("; ") { "${it.type} (${it.reason})" }
            parts += "Ref couldn't apply: $detail"
        }
        return if (parts.isEmpty()) "" else "\n\n${parts.joinToString(" ")}"
    }
}

data class RicoApplyOutcome(
    val root: kotlinx.serialization.json.JsonObject,
    val result: RicoApplyResult,
)

data class RicoCacheApplyOutcome(
    val changed: Boolean,
    val result: RicoApplyResult,
)
